File: src/main/java/rocks/sakira/sakurarosea/common/particle/Particles.java
package rocks.sakira.sakurarosea.common.particle;

import net.minecraft.client.Minecraft;
import net.minecraft.particles.BasicParticleType;
import net.minecraft.particles.ParticleType;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.api.distmarker.OnlyIn;
import net.minecraftforge.event.RegistryEvent;
import net.minecraftforge.eventbus.api.SubscribeEvent;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.registries.ForgeRegistries;
import rocks.sakira.sakurarosea.Constants;

@Mod.EventBusSubscriber(bus = Mod.EventBusSubscriber.Bus.MOD)
public class Particles {
    public static final BasicParticleType SAKURA_LEAF = new BasicParticleType(false);

    @OnlyIn(Dist.CLIENT)
    public static void registerFactories() {
        Minecraft.getInstance().particles.registerFactory(SAKURA_LEAF, SakuraLeafParticle.Factory::new);
    }

    @SubscribeEvent
    public static void registerParticles(RegistryEvent.Register<ParticleType<?>> event) {
        SAKURA_LEAF.setRegistryName(new ResourceLocation(Constants.MOD_ID, "sakura_leaf"));
        ForgeRegistries.PARTICLE_TYPES.register(SAKURA_LEAF);
    }
}

File: src/main/kotlin/io/github/faecraft/sakurarosea/register/Particles.kt
package io.github.faecraft.sakurarosea.register

import net.fabricmc.fabric.api.client.particle.v1.FabricSpriteProvider
import net.fabricmc.fabric.api.client.particle.v1.ParticleFactoryRegistry
import net.fabricmc.fabric.api.particle.v1.FabricParticleTypes
import net.minecraft.client.particle.ParticleFactory
import net.minecraft.particle.DefaultParticleType
import net.minecraft.particle.ParticleType
import net.minecraft.util.Identifier
import net.minecraft.util.registry.Registry
import io.github.faecraft.sakurarosea.MOD_ID
import io.github.faecraft.sakurarosea.tree.SakuraLeafParticle

object Particles {
    val SAKURA_LEAF: DefaultParticleType = FabricParticleTypes.simple()

    fun register() {
        register(SAKURA_LEAF, Identifier(MOD_ID, "sakura_leaf")) { SakuraLeafParticle.Factory(it) }
    }

    fun register(particle: ParticleType<DefaultParticleType>, id: Identifier, factory: (FabricSpriteProvider) -> (ParticleFactory<DefaultParticleType>)) {
        Registry.register(Registry.PARTICLE_TYPE, id, particle)

        ParticleFactoryRegistry.getInstance().register(particle, factory)
    }
}
